Add "named group" feature

kevin-d-omara opened the pull request:
## New Feature
This feature allows the user to optionally associate each Utility Memory Bag with a named group (ex: "Blue Player Factions"). For each named group, only one bag may have it's objects placed at a time. Whenever objects are placed from a bag, it first recalls all other bags in it's named group.

This feature was inspired by the [Memory Bag Groups mod](https://steamcommunity.com/sharedfiles/filedetails/?id=1090295747).

I wrote my own version of the code from scratch and with some improvements:
* The user doesn't need to manually modify the Global script. The Utility Memory Bag takes care of that for you.
* The code for this feature is self contained and object oriented.

## Testing
I've been using this code since August 2019 in my [Advanced Space Crusade mod](https://steamcommunity.com/sharedfiles/filedetails/?id=1848337677). I tested the code extensively at that time, but I don't remember the exact tests. I cleaned up the code a bit for this pull request and did some quick testing for the cleaner code.

## Using the Feature
To associate a Utility Memory Bag with a named group, change the value on line 9 of the script (i.e. `CONFIG.MEMORY_GROUP.NAME`).

I'll add a UI text box for this feature in 1-2 days (see below). That way users don't need scripting skills to use this feature.

## Follow Up
Please don't merge in this pull request yet.

I have a couple more things I want to do. I ran out of time tonight, and will finish this PR within 1-2 days.

Remaining tasks:
* Add UI for this feature (ex: a text box / button).
* Extensive testing. Even though I'm confident with my tests from Aug 2019, I'd like to do my due diligence again to test this feature.

sunflowermans commented:
Thanks! This is looking great so far. I'll keep an eye out for your other changes.

sunflowermans commented on UtilityMemoryBag.lua:
```diff
@@ -1,7 +1,127 @@
+    MEMORY_GROUP = {
+        -- Use an identical value across bags in the group.
+        -- Set this to 'nil' in order to not have the bag in a group.
+        NAME = nil,
```
If you made this a default "nongroup" or something, we could track all Util memory bag guids

sunflowermans commented on UtilityMemoryBag.lua:
```diff
@@ -1,7 +1,127 @@
+
+-- Return "true" if another bag in my memory group has any objects out on the table.
+function anyOtherBagsInMyGroupArePlaced()
+    for bagGuid,_ in pairs(GlobalMemoryGroups:getGroup(CONFIG.MEMORY_GROUP.NAME)) do
```
if using nongroup, would have to check for special case somewhere. Maybe a flag.

sunflowermans reviewed (commented):
Great work here so far. This is not a requirement for this PR, but a thought for the future if you'd like to make a couple more changes...

I'd love to use the global variable to be able to keep track of all Utility Memory Bags guids in the game world (regardless of using the group feature). This would help later if we add an auto-update feature. 

Would only have to click once to update all Utility Memory bags and pull the new code from an online source instead of having to manually update bag scripts.

kevin-d-omara commented:
I love your suggestion!! When I work on this tonight, I'll think about adding that to this PR or doing a follow-up.

kevin-d-omara commented:
I just added another commit to this PR. I implemented the global list containing all Utility Memory Bags in the game world.

I decided to use a separate variable for the global list, since it's independent of the Named Groups feature. This way the code stays simple for these two features.

I'll implement the UI for the Name Groups feature tomorrow. Then we should be good to merge the PR 😊

kevin-d-omara commented:
Done! I added the UI and finished testing. I'm all done on my end.

Here's a Tabletop Simulator save file you can load up to test out the new feature. Just unzip and drop the files into `C:\Users\%USERNAME%\Documents\My Games\Tabletop Simulator\Saves`
[TTS Savefile - Demo.zip](https://github.com/sunflowermans/TTS-UtilityMemoryBag/files/6284336/TTS.Savefile.-.Demo.zip)

sunflowermans commented on UtilityMemoryBag.lua:
```diff
@@ -1,9 +1,226 @@
+    for bagGuid,_ in pairs(GlobalMemoryGroups:getGroup(memoryGroupName:get())) do
+        if bagGuid ~= self.getGUID() then
+            local bag = getObjectFromGUID(bagGuid)
+            local state = bag.call('areAnyOfMyObjectsPlaced')
```
Is this going to cause an error if `bag` is `nil`?

kevin-d-omara replied:
Yes it would. Great catch! I didn't consider that people might delete bags, in which case `GlobalMemoryGroups` would have stale/non-existent guids.

I'll submit a shortly.
